"Won't fix" known issues page filter

gautesl opened the pull request:
Added two new directives to add a filter dropdown to a page. All elements that should be able to be filtered away must contain the "hideable" html class, as well as some class that the filter directive can recognize.

`page-filter` is the main directive which adds generic filtering. When using this directive, all filterable html classes must be given in its body as a tuple (class, displayname). The directive accepts two options:
`name`: Used as a html identifier and to display the "All x" option of the dropdown.
`default`: The default value of the dropdown. Defaults to "all".

`version-filter` is a subclass of the `page-filter` directive which automatically populates the dropdown with all release versions. This directive also adds the hideable class to all elements that already contain the vX-X-X style classes, and creates clickable version tags in the desired container when specifying the `container` option.

Examples of use:
```rst
.. page-filter::
  :name: components

  ble_controller  BLE Controller
  nrfcloud   nRFCloud
  tf-m       TF-M
  ble_mesh   BLE MESH

.. version-filter::
  :default: v2-4-0
  :container: dl/dt
```

This PR also replaces the existing filter function on the known issues page with the new `version-filter` directive, and adds a new "Won't fix" filter.

![image](https://github.com/nrfconnect/sdk-nrf/assets/23337250/70165b94-6a25-4d39-b0cb-2edb1e16015d)

IMPORTANT: These issues are randomly selected for previewing purposes and do NOT reflect actual unresolvable issues!
No wontfix labels are added in this PR, and must be added at a later stage. This means that filtering for "Won't fix" issues currently will produce an empty page.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-11637
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

greg-fer commented:
@gautesl , this looks very nice, and thank you also for simplifying the handling of the long version list.

I have two suggestions:
- Can we have a "Won't fix" tag in red added to the list of version tags whenever we add a `wontfix`?

And a couple of questions:
- What does `dl/dt` stand for?
- Do I understand correctly that `version-filter` can be set to some other value than the latest released version, for example 2.3.0 instead of 2.4.0, yet 2.4.0 will still be available in the list?

@umapraseeda , this needs coordination with @avalei with regard to Zoomin.

gautesl commented:
> @gautesl , this looks very nice, and thank you also for simplifying the handling of the long version list.
> 
> I have a suggestion:
> 
> * Can we have a "Won't fix" tag in red added to the list of version tags whenever we add a `wontfix`?
> 
> And a couple of questions:
> 
> * What does `dl/dt` stand for?
> * Do I understand correctly that `version-filter` can be set to some other value than the latest released version, for example 2.3.0 instead of 2.4.0, yet 2.4.0 will still be available in the list?
> 
> @umapraseeda , this needs coordination with @avalei with regard to Zoomin.

Thank you for suggestions!
I am currently working on a `wontfix` tag, so that will come.

- `dl/dt` is the html node where the labels will be populated into (so that it matches the current system).
- The default version can be any version available from [versions.json](https://github.com/nrfconnect/sdk-nrf/blob/main/doc/versions.json).

greg-fer commented:
> A correct distribution of the wontfix label is a prerequisite for removing the DNM label and upgrading the PR from a draft PR.

@gautesl , can we have the `wontfix` in `page-filter/components` but without populating the tag to the issues? I'm afraid that if we decide to populate known issues with the `wontfix` tag in this PR, it will dilute the purpose of this PR and _significantly_ lengthen its review. I think adding the tag could be handled in the known issues review before each release.

gautesl commented:
Added a visible "Won't fix" tag to to the versiontag list. The tags are created with a new option to the `:page-filter:` directive, i.e. `:page-filter: [("wontfix", "Won't fix")]` will look for html elements with the "wontfix" class and insert tags containing the text "Won't fix".

@greg-fer After the new tag has been deemed satisfactory and the exact wording (now Won't fix - Fixable) has been selected, we can remove the test tags and only insert the dropdown component. It will be a "useless" filter until issues are populated with wontfix-tags (or what they will be called), but that can come later.

umapraseeda commented on doc/nrf/releases_and_maturity/known_issues.rst:
```diff
@@ -21,65 +21,27 @@ A known issue can list one or both of the following entries:
+.. page-filter::
+  :name: issues
+
+  wontfix    Won't fix
+  !wontfix   Fixable
```
@wbober, could you confirm if this wording is good for the filter text? Do you have any other suggestions?

greg-fer replied:
Do we need `Fixable`? AFAIK, all known issues are fixable until tagged with `Won't Fix`, so why do we need a tag for `!wontfix`? It will only confuse the readers, who will now face issues that are fixable, issues that won't be fixed, and issues that have neither of these tags... Not to mention that `Fixable` could change to `Won't Fix` and we would have to explain why this change happened.

gautesl replied:
@greg-fer There is no visible "Fixable" tag, this is only for the dropdown showing the filter options

wbober replied:
> Do we need `Fixable`? AFAIK, all known issues are fixable until tagged with `Won't Fix`, so why do we need a tag for `!wontfix`? It will only confuse the readers, who will now face issues that are fixable, issues that won't be fixed, and issues that have neither of these tags... Not to mention that `Fixable` could change to `Won't Fix` and we would have to explain why this change happened.

I'm with @greg-fer on that. I'd just settle on "All" and "Won't fix" filters.

gautesl replied:
"Fixable" has now been removed as a filter option, and the all wontfix-tags has been removed (a screenshot of how it looks can be found in the PR description instead)

gautesl commented:
Removed all wontfix-tags used for testing and the "Fixable" filtering option. Filtering on "Won't fix" will now yield an empty page. Removed the DNM tag since no false tags are present.

greg-fer commented:
Thank you so much @gautesl for implementing this!

gmarull commented on doc/_extensions/static/page_filter.js:
```diff
@@ -0,0 +1,171 @@
+  dropdown.value = option;
+
+  document.querySelectorAll(".hideable").forEach(e => e.hidden = false);
+  var filters = all_dropdowns
+    .map(dp => dp.options[dp.selectedIndex].value)
+    .filter(val => val !== "all");
+
+  if (filters.length == 0) return;
+
+  var negativeFilters = filters
+    .filter(val => val.charAt(0) === "!")
+    .map(val => val.substring(1));
+  filters = filters.filter(val => val.charAt(0) !== "!")
+
+  let selector = "";
```
nit: this file should likely make consistent usage of var/let

gautesl replied:
Fixed, thanks!

gautesl commented on doc/_extensions/static/page_filter.js:
```diff
@@ -0,0 +1,171 @@
+const all_dropdowns = []
```
This was already treated as a module except for the name, fixed now!

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11637/9/)
- Skipped

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-11637/9/)
- Skipped



*Note: This message is automatically posted and updated by the CI*
